### PR TITLE
RavenDB-12376: librvnpal (continues #8422, #8502)

### PR DESCRIPTION
**Pull request by aviviadi**

### CODE REVIEW

File: src/Voron/Platform/Pal.cs [hunk 1]
```diff
@@ -46,7 +46,7 @@ static Pal()
 
             try
             {
-                if (File.Exists(toFilename) == false)
+                if (File.Exists(fromFilename))
                     File.Move(fromFilename, toFilename);
```
Comment (ayende) [inline src/Voron/Platform/Pal.cs]: Why did you made this change?

Comment (aviviadi) [reply src/Voron/Platform/Pal.cs]: because of https://issues.hibernatingrhinos.com/issue/RavenDB-12660#focus=streamItem-67-33793-0-0

If someone copied new lib, we will move it to the general name. This way we avoid unnecessary copy tries and do it once per publish

Comment (ayende) [reply src/Voron/Platform/Pal.cs]: I see, okay

**Review by ppekrol — APPROVED** [review]

File: src/Voron/Exceptions/DiskFullException.cs [hunk 2]
```diff
@@ -18,10 +18,10 @@ public DiskFullException()
         public string DirectoryPath;
         public long CurrentFreeSpace;
 
-        public DiskFullException(string filePath, long requestedFileSize, long? freeSpace)
+        public DiskFullException(string filePath, long requestedFileSize, long? freeSpace, string msg)
             : base(
                 $"There is not enough space to set the size of file {filePath} to {Sizes.Humane(requestedFileSize)}. " +
-                $"Currently available space: {Sizes.Humane(freeSpace) ?? "N/A"}"
+                $"Currently available space: {Sizes.Humane(freeSpace) ?? "N/A"}. Error Message:{msg}"
```
Comment (ayende) [inline src/Voron/Exceptions/DiskFullException.cs]:
```suggestion
                $"Currently available space: {Sizes.Humane(freeSpace) ?? "N/A"}. Error Message: {msg}"
```

File: src/Voron/Impl/Journal/IJournalWriter.cs [hunk 3]
```diff
@@ -1,22 +1,217 @@
...
+        void Read(byte* buffer, long numOfBytes, long offsetInFile);
         void Truncate(long size);
 
         void AddRef();
         bool Release();
 
         VoronPathSetting FileName { get; }
     }
+
+    public unsafe class JournalWriter : IJournalWriter
```
Comment (ayende) [inline src/Voron/Impl/Journal/IJournalWriter.cs]: Let's move to a separate file

File: src/Voron/Impl/Journal/IJournalWriter.cs [hunk 4]
```diff
@@ -1,22 +1,217 @@
...
+
+        public void Write(long posBy4Kb, byte* p, int numberOf4Kb)
+        {
+            Debug.Assert(_options.IoMetrics != null);
+
+            using (var metrics = _options.IoMetrics.MeterIoRate(FileName.FullPath, IoMetrics.MeterType.JournalWrite, (long)numberOf4Kb * 4 * Constants.Size.Kilobyte))
+            {
+                var result = Pal.rvn_write_journal(_writeHandle, p, numberOf4Kb * 4 * Constants.Size.Kilobyte, posBy4Kb * 4 * Constants.Size.Kilobyte, out var error);
+                if (result != PalFlags.FailCodes.Success)
+                    PalHelper.ThrowLastError(result, error, $"Attempted to write to journal file - Path:{FileName.FullPath} Size:{numberOf4Kb * 4 * Constants.Size.Kilobyte}");
```
Comment (ayende) [inline src/Voron/Impl/Journal/IJournalWriter.cs]:
```suggestion
                    PalHelper.ThrowLastError(result, error, $"Attempted to write to journal file - Path: {FileName.FullPath} Size: {numberOf4Kb * 4 * Constants.Size.Kilobyte}");
```

File: src/Voron/Impl/Journal/IJournalWriter.cs [hunk 5]
```diff
@@ -1,22 +1,217 @@
...
+                out actualSize,
+                out errorCode
+                );
+
+            EnsureValidResult();
+            
+            void EnsureValidResult()
+            {
+                if (result != PalFlags.FailCodes.Success)
+                    PalHelper.ThrowLastError(result, errorCode, $"Attempted to read from journal file - Path:{FileName.FullPath} Size:{numOfBytes} Offset:{offsetInFile} ActualSize:{actualSize}");
```
Comment (ayende) [inline src/Voron/Impl/Journal/IJournalWriter.cs]:
```suggestion
                    PalHelper.ThrowLastError(result, errorCode, $"Attempted to read from journal file - Path: {FileName.FullPath} Size: {numOfBytes} Offset: {offsetInFile} ActualSize: {actualSize}");
```

File: src/Voron/Impl/Journal/IJournalWriter.cs [hunk 6]
```diff
@@ -1,22 +1,217 @@
...
+
+            }
+        }
+
+        public void Truncate(long size)
+        {
+            var result = Pal.rvn_truncate_journal(_writeHandle, size, out var error);
+            if (result != PalFlags.FailCodes.Success)
+                PalHelper.ThrowLastError(result, error, $"Attempted to write to journal file - Path:{FileName.FullPath} Size:{size}");
```
Comment (ayende) [inline src/Voron/Impl/Journal/IJournalWriter.cs]:
```suggestion
                PalHelper.ThrowLastError(result, error, $"Attempted to write to journal file - Path: {FileName.FullPath} Size: {size}");
```

File: src/Voron/Impl/Journal/IJournalWriter.cs [hunk 7]
```diff
@@ -1,22 +1,217 @@
...
+            _options.IoMetrics.FileClosed(FileName.FullPath);
+            
+            List<Exception> exceptions = null;
+            
+            TryExecute(() =>
+            {
+                _readHandle.Dispose();
+                if (_readHandle.FailCode != PalFlags.FailCodes.Success)
+                    PalHelper.ThrowLastError(_readHandle.FailCode, _readHandle.ErrorNo,
+                        $"Attempted to close 'read journal handle' - Path:{FileName.FullPath}");
```
Comment (ayende) [inline src/Voron/Impl/Journal/IJournalWriter.cs]:
```suggestion
                        $"Attempted to close 'read journal handle' - Path: {FileName.FullPath}");
```

File: src/Voron/Impl/Journal/IJournalWriter.cs [hunk 8]
```diff
@@ -1,22 +1,217 @@
...
+            });
+
+            TryExecute(() =>
+            {
+                _writeHandle.Dispose();
+                if (_writeHandle.FailCode != PalFlags.FailCodes.Success)
+                    PalHelper.ThrowLastError(_writeHandle.FailCode, _writeHandle.ErrorNo,
+                        $"Attempted to close 'write journal handle' - Path:{FileName.FullPath}");
```
Comment (ayende) [inline src/Voron/Impl/Journal/IJournalWriter.cs]:
```suggestion
                        $"Attempted to close 'write journal handle' - Path: {FileName.FullPath}");
```

File: src/Voron/Impl/Paging/AbstractPager.cs [hunk 9]
```diff
@@ -601,48 +603,26 @@ public virtual void DiscardWholeFile()
...
-                    NumberOfBytes = new IntPtr(size)
-                };
-
-                size = 0;
-                baseAddress = null;
-            }
-            PrefetchRanges(ranges, pagerState.AllocationInfos.Length);
+            var rc = Pal.rvn_prefetch_virtual_memory(allocationInfo.BaseAddress, allocationInfo.Size, out var errorCode);
+            if(rc != PalFlags.FailCodes.Success)
+                Log.Info(PalHelper.GetNativeErrorString(errorCode, $"Tried to prefetch whole file. Result:{rc} FileName:${FileName.FullPath} Size:{allocationInfo.Size}", out _));
```
Comment (ayende) [inline src/Voron/Impl/Paging/AbstractPager.cs]:
```suggestion
                Log.Info(PalHelper.GetNativeErrorString(errorCode, $"Tried to prefetch whole file. Result: {rc} FileName: {FileName.FullPath} Size: {allocationInfo.Size}", out _));
```

File: src/Voron/Impl/Journal/IJournalWriter.cs [hunk 10]
```diff
@@ -1,22 +1,217 @@
...
+                metrics.SetFileSize(NumberOfAllocated4Kb * (4L * Constants.Size.Kilobyte));
+            }
+        }
+
+        public AbstractPager CreatePager()
+        {
+            if (_options.RunningOn32Bits)
+                return new Windows32BitsMemoryMapPager(_options, FileName);
+
+            return new WindowsMemoryMapPager(_options, FileName);
```
Comment (ayende) [inline src/Voron/Impl/Journal/IJournalWriter.cs]: Shouldn't you use `RvnMemoryMapPager` here?

File: src/Voron/Impl/Paging/RvnMemoryMapPager.cs [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+                out _handle,
+                out var baseAddress,
+                out _totalAllocationSize,
+                out var errorCode);
+
+            if (rc != FailCodes.Success)
+            {
+                try
+                {
+                    PalHelper.ThrowLastError(rc, errorCode, $"rvn_create_and_mmap64_file failed on {rc}");
```
Comment (ayende) [inline src/Voron/Impl/Paging/RvnMemoryMapPager.cs]: Where is the file name here?

File: src/Voron/Impl/Paging/RvnMemoryMapPager.cs [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+            try
+            {
+                using (var metric = Options.IoMetrics.MeterIoRate(FileName.FullPath, IoMetrics.MeterType.DataSync, 0))
+                {
+                    foreach (var alloc in currentState.AllocationInfos)
+                    {
+                        metric.IncrementFileSize(alloc.Size);
+                        var rc = rvn_memory_sync(alloc.BaseAddress, alloc.Size, out var errorCode);
+                        if (rc != FailCodes.Success)
+                            PalHelper.ThrowLastError(rc, errorCode,$"Failed to memory sync at ${new IntPtr(alloc.BaseAddress).ToInt64():X}. TotalUnsynced = ${totalUnsynced}");
```
Comment (ayende) [inline src/Voron/Impl/Paging/RvnMemoryMapPager.cs]: Where is the file name here?

File: src/Voron/Impl/Paging/RvnMemoryMapPager.cs [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+            NativeMemory.UnregisterFileMapping(FileName?.FullPath, new IntPtr(baseAddress), size);
+        }
+
+        protected override void DisposeInternal()
+        {
+            _handle.Dispose();
+            // _handle.FailCode != success, we cannot delete the file probably, and there's nothing much we can do here.
+            // just add to log and continue            
+            if (_handle.FailCode != FailCodes.Success)
+                if (_logger.IsInfoEnabled)
```
Comment (ayende) [inline src/Voron/Impl/Paging/RvnMemoryMapPager.cs]: Please add braces here

File: src/Voron/Impl/Paging/RvnMemoryMapPager.cs [hunk 14]
```diff
@@ -0,0 +1,251 @@
...
+
+            var newLengthAfterAdjustment = NearestSizeToPageSize(newLength);
+
+            if (newLengthAfterAdjustment <= _totalAllocationSize)
+                return null;
+
+            var rc = rvn_allocate_more_space(newLengthAfterAdjustment, _handle, out var newAddress, out var errorCode);
+
+            if (rc != FailCodes.Success)
+                PalHelper.ThrowLastError(rc, errorCode, $"can't allocate more pages (rc={rc}). Requested {newLength} (adjusted to {newLengthAfterAdjustment})");
```
Comment (ayende) [inline src/Voron/Impl/Paging/RvnMemoryMapPager.cs]: Where is the file name?

File: src/Voron/Impl/Paging/RvnMemoryMapPager.cs [hunk 15]
```diff
@@ -0,0 +1,251 @@
...
+        {
+            if (size == 0 || 
+                UsePageProtection == false && force == false) 
+                return;
+
+            if (rvn_protect_range(start, (long)size, ProtectRange.Protect, out var errorCode) == 0)
+                return;
+
+            if (_logger.IsInfoEnabled)
+                _logger.Info($"Unable to protect page range. start={new IntPtr(start).ToInt64():X}, size={size}, PROT_READ, errorCode={errorCode}");
```
Comment (ayende) [inline src/Voron/Impl/Paging/RvnMemoryMapPager.cs]: What is the file name here?

File: src/Voron/Impl/Paging/RvnMemoryMapPager.cs [hunk 16]
```diff
@@ -0,0 +1,251 @@
...
+            Debugger.Break();
```
Comment (ayende) [inline src/Voron/Impl/Paging/RvnMemoryMapPager.cs]: Need to remove this.

File: src/Voron/Impl/Paging/RvnMemoryMapPager.cs [hunk 17]
```diff
@@ -0,0 +1,251 @@
...
+            if (size == 0 || 
+                UsePageProtection == false && force == false) 
+                return;
+
+            if (rvn_protect_range(start, (long)size, ProtectRange.Unprotect, out var errorCode) == 0)
+                return;
+
+            if (_logger.IsInfoEnabled)
+                _logger.Info($"Unable to un-protect page range. start={new IntPtr(start).ToInt64():X}, size={size}, PROT_READ | PROT_WRITE, errorCode={errorCode}");
+            Debugger.Break();
```
Comment (ayende) [inline src/Voron/Impl/Paging/RvnMemoryMapPager.cs]: Need to remove this

File: src/Voron/Impl/Paging/RvnMemoryMapPager.cs [hunk 18]
```diff
@@ -0,0 +1,251 @@
...
+        {
+            if (size == 0 || 
+                UsePageProtection == false && force == false) 
+                return;
+
+            if (rvn_protect_range(start, (long)size, ProtectRange.Unprotect, out var errorCode) == 0)
+                return;
+
+            if (_logger.IsInfoEnabled)
+                _logger.Info($"Unable to un-protect page range. start={new IntPtr(start).ToInt64():X}, size={size}, PROT_READ | PROT_WRITE, errorCode={errorCode}");
```
Comment (ayende) [inline src/Voron/Impl/Paging/RvnMemoryMapPager.cs]: Where is the file name here?
Also, PROT_READ is linux, so not right.

File: src/Voron/Impl/Paging/RvnMemoryMapPager.cs (hunk 15 above)
Comment (ayende) [reply src/Voron/Impl/Paging/RvnMemoryMapPager.cs]: PROT_READ is not cross platform

File: src/Voron/Platform/PalDefinitions.cs [hunk 19]
```diff
@@ -0,0 +1,25 @@
+using System;
+using System.Runtime.InteropServices;
+using Voron.Global;
+
+namespace Voron.Platform
+{
+    public unsafe class PalDefinitions
+    {
+        public const int AllocationGranularity = 64 * Constants.Size.Kilobyte;
+        public struct SystemInformation
```
Comment (ayende) [inline src/Voron/Platform/PalDefinitions.cs:12]:
```suggestion
        [StructLayout(LayoutKind.Sequential, Pack = 1)]
        public struct SystemInformation
```

File: src/librvnpal/src/posix/fileutils.c [hunk 20]
```diff
@@ -0,0 +1,191 @@
...
+#include <sys/mman.h>
+#include <assert.h>
+#include <time.h>
+
+#include "rvn.h"
+#include "status_codes.h"
+#include "internal_posix.h"
+
+PRIVATE int64_t
+_nearest_size_to_page_size(int64_t orig_size, int64_t sys_page_size)
```
Comment (ayende) [inline src/librvnpal/src/posix/fileutils.c]: Isn't this common to all platforms?

File: src/librvnpal/src/posix/fileutils.c [hunk 21]
```diff
@@ -0,0 +1,191 @@
...
+
+PRIVATE int32_t
+_ensure_path_exists(const char* path)
+{
+    /* TODO: implement */
+    return SUCCESS;   
+}
+
+PRIVATE int32_t 
+_open_file_to_read(const char *file_name, void **handle, int32_t *detailed_error_code)
```
Comment (ayende) [inline src/librvnpal/src/posix/fileutils.c]: In open for write, we are using `journal_handle`, here we are storing the raw fd in the void*.
We need to be consistent, because otherwise, it can get messy

File: src/librvnpal/src/posix/journal.c [hunk 22]
```diff
@@ -0,0 +1,195 @@
...
+    struct stat fs;
+    if (fstat(jfh->fd, &fs) == -1)
+    {
+        rc = FAIL_STAT_FILE;
+        goto error_cleanup;
+    }
+
+    if (fs.st_size < initial_file_size)
+    {
+        rc = _resize_file((void*)(int64_t)jfh->fd, initial_file_size, detailed_error_code);
```
Comment (ayende) [inline src/librvnpal/src/posix/journal.c]:
```suggestion
        rc = _resize_file(jfh->fd, initial_file_size, detailed_error_code);
```

File: src/librvnpal/src/posix/fileutils.c (hunk 21 above)
Comment (ayende) [reply src/librvnpal/src/posix/fileutils.c]: This is an internal impl detail, use it as such 
```suggestion
_open_file_to_read(const char *file_name, int *fd, int32_t *detailed_error_code)
```

File: src/librvnpal/src/posix/fileutils.c [hunk 23]
```diff
@@ -0,0 +1,191 @@
...
+        *handle = (void*)(int64_t)fd;
+        return SUCCESS;
+    }
+
+    *detailed_error_code = errno;
+    return FAIL_OPEN_FILE;
+}
+
+PRIVATE int32_t
+_read_file(void *handle, void *buffer, int64_t required_size, int64_t offset, int64_t *actual_size, int32_t *detailed_error_code)
```
Comment (ayende) [inline src/librvnpal/src/posix/fileutils.c]:
```suggestion
_read_file(int fd, void *buffer, int64_t required_size, int64_t offset, int64_t *actual_size, int32_t *detailed_error_code)
```

File: src/librvnpal/src/posix/journal.c [hunk 24]
```diff
@@ -0,0 +1,195 @@
...
+#include "status_codes.h"
+#include "internal_posix.h"
+
+struct journal_handle
+{
+    int fd;
+    const char *path;
+};
+
+PRIVATE void free_journal_handle(struct journal_handle* handle)
```
Comment (ayende) [inline src/librvnpal/src/posix/journal.c]:
```suggestion
PRIVATE void _free_journal_handle(struct journal_handle* handle)
```

Because it is internal, no?

File: src/librvnpal/src/posix/journal.c [hunk 25]
```diff
@@ -0,0 +1,195 @@
...
+    struct journal_handle *jfh = calloc(1, sizeof(struct journal_handle));
+    *handle = jfh;
+    if (jfh == NULL)
+    {
+        *detailed_error_code = errno;
+        return FAIL_CALLOC;
+    }
+
+    jfh->path = NULL;
+    rc = _open_file_to_read(file_name, (void**)&(jfh->fd), detailed_error_code);
```
Comment (ayende) [inline src/librvnpal/src/posix/journal.c]:
```suggestion
    rc = _open_file_to_read(file_name, &jfh->fd, detailed_error_code);
```

File: src/librvnpal/src/posix/journal.c [hunk 26]
```diff
@@ -0,0 +1,195 @@
...
+    }
+
+    return rc;
+}
+
+EXPORT int32_t
+rvn_read_journal(void *handle, void *buffer, int64_t required_size, int64_t offset, int64_t *actual_size, int32_t *detailed_error_code)
+{
+    struct journal_handle *jfh = (struct journal_handle *)handle;
+    return _read_file((void*)(int64_t)(jfh->fd), buffer, required_size, offset, actual_size, detailed_error_code);
```
Comment (ayende) [inline src/librvnpal/src/posix/journal.c]:
```suggestion
    return _read_file(jfh->fd, buffer, required_size, offset, actual_size, detailed_error_code);
```

File: src/librvnpal/src/posix/journal.c [hunk 27]
```diff
@@ -0,0 +1,195 @@
...
+    int32_t rc;
+    struct journal_handle *jfh = (struct journal_handle *)handle;
+
+    if (_flush_file(jfh->fd) == -1)
+    {
+        rc = FAIL_SYNC_FILE;
+        goto error_cleanup;
+    }
+
+    rc = _resize_file((void *)(int64_t)(jfh->fd), size, detailed_error_code);
```
Comment (ayende) [inline src/librvnpal/src/posix/journal.c]:
```suggestion
    rc = _resize_file(jfh->fd, size, detailed_error_code);
```

File: src/librvnpal/src/posix/mapping.c [hunk 28]
```diff
@@ -0,0 +1,249 @@
...
+    }
+    mfh->flags = flags;
+    mfh->fd = open(path, O_RDWR | O_CREAT, S_IWUSR | S_IRUSR);
+    if (mfh->fd == -1)
+    {
+        rc = FAIL_OPEN_FILE;
+        goto error_cleanup;
+    }
+
+    int64_t sz = lseek(mfh->fd, 0, SEEK_END);
```
Comment (ayende) [inline src/librvnpal/src/posix/mapping.c]: Minor, `fstat` might be better here

File: src/librvnpal/src/posix/mapping.c [hunk 29]
```diff
@@ -0,0 +1,249 @@
...
+    {
+        rc = FAIL_SYSCONF;
+        goto error_cleanup;
+    }
+
+    int32_t allocation_granularity = ALLOCATION_GRANULARITY;
+    if (initial_file_size < allocation_granularity)
+        initial_file_size = allocation_granularity;
+
+    if (sz <= initial_file_size || sz % sys_page_size != 0)
```
Comment (ayende) [inline src/librvnpal/src/posix/mapping.c:97]: Here you are _always_ syncing the directory on opening a file, why?
We only need to do that if the file size changed.

Comment (ayende) [reply src/librvnpal/src/posix/mapping.c:97]: Or the file was created

File: src/librvnpal/src/posix/journal.c [hunk 30]
```diff
@@ -0,0 +1,195 @@
...
+    
+    int32_t flags = O_DSYNC | O_DIRECT;
+    if (transaction_mode == JOURNAL_MODE_DANGER)
+        flags = 0;
+
+    if (sizeof(int) == 4) /* 32 bits */
+        flags |= O_LARGEFILE;
+
+    struct journal_handle *jfh = calloc(1, sizeof(struct journal_handle));
+    *handle = jfh;
```
Comment (ayende) [inline src/librvnpal/src/posix/journal.c:53]: Here need to set the flags as delete on close if use pure memory mode

File: test/FastTests/Voron/FixedSize/LargeFixedSizeTreeBugs.cs [hunk 31]
```diff
@@ -20,6 +20,8 @@ protected override void Configure(StorageEnvironmentOptions options)
         [Fact]
         public void DeleteRangeShouldModifyPage()
         {
+            RequireFileBasedPager();
```
Comment (ayende) [inline test/FastTests/Voron/FixedSize/LargeFixedSizeTreeBugs.cs]: Why is this required?

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Voron/Impl/Journal/IJournalWriter.cs (hunk 10 above)
Comment (aviviadi) [reply src/Voron/Impl/Journal/IJournalWriter.cs]: yes... FastTest / Regular import was able to run without it (pure mem etc.)

File: src/librvnpal/src/posix/fileutils.c (hunk 21 above)
Comment (aviviadi) [reply src/librvnpal/src/posix/fileutils.c]: No. This is (will, when windows' map pager will be pal implemented) need to fit for both platforms. So the header must be **handle. Unless we want different headers (_open_file_to_read_linux and _open_file_to_read_windows, and call the appropriate from the posix/win rvn_* func)

File: src/librvnpal/src/posix/journal.c (hunk 22 above)
Comment (aviviadi) [reply src/librvnpal/src/posix/journal.c]: Why? (this prevents a compiler warning)

File: src/librvnpal/src/posix/fileutils.c (hunk 23 above)
Comment (aviviadi) [reply src/librvnpal/src/posix/fileutils.c]: same as in https://github.com/ravendb/ravendb/pull/8552#discussion_r251450373

File: src/librvnpal/src/posix/journal.c (hunk 25 above)
Comment (aviviadi) [reply src/librvnpal/src/posix/journal.c]: same as in https://github.com/ravendb/ravendb/pull/8552#discussion_r251450899

File: src/librvnpal/src/posix/journal.c (hunk 26 above)
Comment (aviviadi) [reply src/librvnpal/src/posix/journal.c]: same as in https://github.com/ravendb/ravendb/pull/8552#discussion_r251450899

File: src/librvnpal/src/posix/journal.c (hunk 27 above)
Comment (aviviadi) [reply src/librvnpal/src/posix/journal.c]: https://github.com/ravendb/ravendb/pull/8552#discussion_r251450899

File: src/librvnpal/src/posix/journal.c (hunk 22 above)
Comment (ayende) [reply src/librvnpal/src/posix/journal.c]: The sig for `_resize_file` should be `_resize_file(int fd, ...)`

File: src/librvnpal/src/posix/fileutils.c (hunk 21 above)
Comment (ayende) [reply src/librvnpal/src/posix/fileutils.c]: This is private for the impl. No need to expose the xplat stuff here.

File: src/Voron/Impl/Journal/JournalWriter.cs [hunk 32]
```diff
@@ -0,0 +1,191 @@
...
+                if (_options.RunningOn32Bits)
+                    return new Posix32BitsMemoryMapPager(_options, FileName);
+
+                return new RvnMemoryMapPager(_options, FileName);
+            }
+
+            if (_options.RunningOn32Bits)
+                return new Windows32BitsMemoryMapPager(_options, FileName);
+
+            return new WindowsMemoryMapPager(_options, FileName);
```
Comment (ayende) [inline src/Voron/Impl/Journal/JournalWriter.cs:70]: Why not `RvnMemoryMapPager` here?

Comment (ayende) [reply src/Voron/Impl/Journal/JournalWriter.cs:70]: For that matter, why isn't `WindowsMemoryMapPager` deleted?

File: src/librvnpal/inc/internal_posix.h [hunk 33]
```diff
@@ -47,6 +47,14 @@ _sync_directory_maybe_symblink(char *dir_path, int32_t depth, int32_t *detailed_
 PRIVATE int32_t
 _allocate_file_space(int32_t fd, int64_t size, int32_t *detailed_error_code);
 
+PRIVATE int32_t
+_open_file_to_read(const char* file_name, int32_t *fd, int32_t* detailed_error_code);
+
+PRIVATE int32_t
+_resize_file(int32_t *fd, int64_t size, int32_t *detailed_error_code);
```
Comment (ayende) [inline src/librvnpal/inc/internal_posix.h]: `int32_t fd`

File: src/librvnpal/inc/internal_posix.h [hunk 34]
```diff
@@ -47,6 +47,14 @@ _sync_directory_maybe_symblink(char *dir_path, int32_t depth, int32_t *detailed_
...
+
+PRIVATE int32_t
+_read_file(int32_t *fd, void* buffer, int64_t required_size, int64_t offset, int64_t* actual_size, int32_t* detailed_error_code);
```
Comment (ayende) [inline src/librvnpal/inc/internal_posix.h]: `int32_t fd`

File: src/librvnpal/inc/internal_win.h [hunk 35]
```diff
@@ -7,7 +7,16 @@ PRIVATE int32_t
...
 
 PRIVATE int32_t
-write_file_in_sections(void* handle, const char* buffer, int64_t size, int64_t offset, uint32_t section_size, int32_t* detailed_error_code);
+_write_file_in_sections(void* handle, const char* buffer, int64_t size, int64_t offset, uint32_t section_size, int32_t* detailed_error_code);
+
+PRIVATE int32_t
+_open_file_to_read(const char *file_name, HANDLE *handle, int32_t *detailed_error_code);
+
+PRIVATE int32_t
+_resize_file(HANDLE *handle, int64_t size, int32_t *detailed_error_code);
```
Comment (ayende) [inline src/librvnpal/inc/internal_win.h]: `HANDLE *handle`

File: src/librvnpal/inc/internal_win.h [hunk 36]
```diff
@@ -7,7 +7,16 @@ PRIVATE int32_t
...
+
+PRIVATE int32_t
+_read_file(HANDLE *handle, void* buffer, int64_t required_size, int64_t offset, int64_t* actual_size, int32_t* detailed_error_code);
```
Comment (ayende) [inline src/librvnpal/inc/internal_win.h]: `HANDLE *handle`

File: src/librvnpal/src/posix/journal.c [hunk 37]
```diff
@@ -42,18 +44,23 @@ rvn_open_journal_for_writes(const char *file_name, int32_t transaction_mode, int
...
+    
     *handle = jfh;
     if (jfh == NULL)
     {
         rc = FAIL_CALLOC;
         goto error_cleanup;
     }
+
+    jfh->delete_on_close = false;
+    if (transaction_mode == JOURNAL_MODE_PURE_MEMORY)
```
Comment (ayende) [inline src/librvnpal/src/posix/journal.c]: `jfh->delete_on_close = transaction_mode == JOURNAL_MODE_PURE_MEMORY;`

**Review by ayende — CHANGES_REQUESTED** [review]

**Review by ayende — APPROVED** [review]
Rebuild and let me know it is ready for merge

**Comment by ayende** [comment]
At least this test seems suspicious: can you check?

SlowTests.Voron.Bugs.RavenDB_6971.Applying_new_diff_requires_to_zero_destination_bytes_first
